src/protocols/MeterFile.cpp

```diff
@@ -128,16 +165,63 @@ int MeterFile::open() {
 
 int MeterFile::close() {
 
+	if (_notify_fd != -1) {
+		(void)::close(_notify_fd);
+		_notify_fd = -1;
+	}
+
 	return fclose(_fd);
 }
 
 ssize_t MeterFile::read(std::vector<Reading> &rds, size_t n) {
 
-	// TODO use inotify to block reading until file changes
-
 	char line[256], *endptr;
 	char *string=0;
 
+	// wait for file change via inotify
+	const int EVENTSIZE = sizeof(struct inotify_event) + NAME_MAX + 1;
+	if (_notify_fd!=-1){
+		// read all events from fd:
+		char buf[EVENTSIZE];
+		ssize_t len;
+
+		int nr_events = 0;
+		do{
+			int totalPending = 0;
+			if (nr_events) {
+				// no blocking expected on 2nd call
+				(void)ioctl(_notify_fd, FIONREAD, &totalPending);
+			}
+
+			if (nr_events==0 || totalPending>0) {
+				len = ::read(_notify_fd, buf, sizeof(buf));	// read will block until inotify event occurs
+				if (len>0)
+					nr_events++;
+			}
+			else
+				len = 0;
+
+			const struct inotify_event *event = (struct inotify_event *)(&buf[0]);
```

Comment:
this code should be moved inside the if (nr_event...) brackets otherwise for len=0 case it might behave random.